### ISSUE TITLE
Block Key adds +1 to count when recalculated

### ISSUE THREAD

**Issue by Tyguh** [BG-Software-LLC/SuperiorSkyblock2#483]
### Minecraft's Version

Paper 1.8.8

### Plugin's Version

B238

### Describe the bug

Whenever the value of an island is recalculated whether it's by using /is recalc or /is admin recalc, it adds +1 to my custom block key counts which also results in the worth of the island being increased. I've tested with the latest build of ss2 but the issue is persistent.

### To Reproduce

Essentially you just place down the block(s) registered as a custom block key, note the count using /is counts. You then run either /is recalc or /is admin recalc and check your /is counts again noticing it increases by 1.

### Additional Information

Before Recalculating: https://i.imgur.com/RrvrCvU.png

After Recalculation: https://i.imgur.com/naLeTkP.png

**Comment by OmerBenGera** [comment]
Fixed in latest dev build

**Comment by OmerBenGera** [comment]
Are you sure you use b238, as it should be fixed in this version

**Comment by OmerBenGera** [comment]
also, spawners are not a custom block key, so I don't understand what you talk about..
did you register them as a custom block key?

**Comment by Tyguh** [comment]
> Are you sure you use b238, as it should be fixed in this version
Yes, I'm positive. https://i.imgur.com/JPvlogz.png


> also, spawners are not a custom block key, so I don't understand what you talk about..
> did you register them as a custom block key?
Yeah, they were registered as custom block keys to accommodate upgrades.

**Comment by OmerBenGera** [comment]
I will look into that.

**Comment by OmerBenGera** [comment]
Mind sharing with me how you registered your custom keys?
I tested it with a custom key I registered (taken from the wiki), and it seems to work fine for me.
I had 4 placed, then I run /is admin recalc, and it stayed at 4.

**Comment by Tyguh** [comment]
Sure thing,
- https://pastebin.com/QAs6KZDn
- https://pastebin.com/DNwJRzvG

**Comment by OmerBenGera** [comment]
You are registering a few key parsers that do the same logic for the same block.
Also, calling Key.of(Block) in the key parsing may lead to infinite-recursive call, so take this in mind.
I recommend cleaning the code and follow the tutorial: [here](https://wiki.bg-software.com/#/superiorskyblock/api/?id=register-your-own-block-keys).
Let me know if it still occurs after you fix that issue.

**Comment by Tyguh** [comment]
Hey sorry, I sent you outdated code here's the updated code.

- https://pastebin.com/LmsSLx1F
- https://pastebin.com/Nrf6H1Um

**Comment by OmerBenGera** [comment]
You still have multiple stack keys with the same logic on the same block (on a spawner block).
Your idea of having a different "StackKey" for each upgrade cannot work with the system of keys.
There should be one key for a spawner, and in this key all the logic should be - checks for upgrades, etc.

**Comment by OmerBenGera** [comment]
```
   @Override
    public Key getCustomKey(Location location) {
        StackedSpawner stackedSpawner = WildStackerAPI.getWildStacker().getSystemManager().getStackedSpawner(location);
        EntityUpgrade entityUpgrade = SuperiorSkyblockHook.getEntityUpgrade(stackedSpawner.getSpawnedType(), stackedSpawner.getUpgrade());
        return entityUpgrade != null ? Key.of(SuperiorSkyblockHook.upgradesKeys.get(entityUpgrade)) : null;
    }
```
Something like this should work, and it should be registered only once.

**Comment by JamesSmily** [comment]
Jesus died cause of the fucking scum bag jews that killed him and i'm glad hitler murdered those scum bags like omre and his family. Boy got mad cause i said jesus got killed by the jews which he is also jewish and banned me from the discord and that is why the jews got killed cause of being scum like him. No money poor af living in a poor country relying on minecraft for money but even that didn't work so he quit cause he's depressed.

**Comment by JamesSmily** [comment]
Jesus died cause of the fucking scum bag jews that killed him and i'm glad hitler murdered those scum bags like omre and his family. Boy got mad cause i said jesus got killed by the jews which he is also jewish and banned me from the discord and that is why the jews got killed cause of being scum like him. No money poor af living in a poor country relying on minecraft for money but even that didn't work so he quit cause he's depressed.

**Comment by JamesSmily** [comment]
Jesus died cause of the fucking scum bag jews that killed him and i'm glad hitler murdered those scum bags like omre and his family. Boy got mad cause i said jesus got killed by the jews which he is also jewish and banned me from the discord and that is why the jews got killed cause of being scum like him. No money poor af living in a poor country relying on minecraft for money but even that didn't work so he quit cause he's depressed.

**Comment by OmerBenGera** [comment]
Any news regarding this?

**Comment by OmerBenGera** [comment]
> > Any news regarding this?
> 
> Hello Omer, I've been asked to take a look at this problem to find out the problem.
> My conclusion is that it is a problem with _**SuperiorSkyblock2**_ and not the plugin.
> How I tested:
> 
> 1. Completely new server of 1.8.8 created from BuildTools.
> 2. Remove everything inside the plugin except for a single key registration for a single type of mob (Creeper).
> 3. Only the 3 necessary plugins present - _**SuperiorSkyblock2**_, _**WildStacker**_, and this plugin.
>    As you can guess for yourself the problem still occurred.
> 
> I'm not familiar with the inner workings of _**SuperiorSkyblock2**_, but my guess is that it calculates the block itself on top of the keys when using custom keys.
> מקווה שיסתדר

What is the spawner-provider set in your config?
I think the issue is that you register custom keys for spawners alongside of WS hook, and therefore you have duplication.

**Comment by Tyguh** [comment]
The `spawners-provider` is set to AUTO

**Comment by OmerBenGera** [comment]
> 
> 
> The `spawners-provider` is set to AUTO

Try setting it to a random, "BLABLBAL" will do the work
Let me know if it still occurs after that (make sure you restart the server)

**Comment by DMan1629** [comment]
Even after setting both "spawners-provider" and "stacked-blocks-provider" to "WildStacker" and later on to "BLABLBAL" the problem still occurs.

**Comment by OmerBenGera** [comment]
one last question - does it happen without WildStacker installed?

**Comment by DMan1629** [comment]
Uhm... of course not... The plugin depends on WildStacker

**Comment by DMan1629** [comment]
But without this plugin _**SuperiorSkyblock2**_ doesn't even recognize the spawner (even after recalculating)

**Comment by OmerBenGera** [comment]
Ah I understand.
Well, I'll try creating a custom key for spawners and get back to you soon.
I will again follow the tutorial from the wiki and get back to you.

Btw, why are you overriding spawners when you have WildStacker installed?
Or even better, why aren't you using the spawners provider for spawners?

**Comment by DMan1629** [comment]
I tried setting the "spawners-provider" and "stacked-blocks-provider" to WildStacker without this plugin but it did nothing.

**Comment by Tyguh** [comment]
> Btw, why are you overriding spawners when you have WildStacker installed?
> Or even better, why aren't you using the spawners provider for spawners?

To allow for different worths per upgrade

**Comment by OmerBenGera** [comment]
Ok, it seems to work fine for me, but I have some points to make:
1) Make sure when you register your keys, you set them with the following format: `SPAWNER:<CUSTOM_TYPE>`
Do not use more than one `:`, otherwise it can break things.
2) Make sure your isCustomKey is overridden and working correctly, otherwise things will be messed up.
3) Make sure you register your custom prices via the worth file.

So I made a small plugin that takes each spawner and registers a custom key for it like this: `SPAWNER:<UPGRADE>_<ENTITY-TYPE>`
I also took the default upgrades and applied them to cows, and that resulted 5 new keys for cow spawners:
`SPAWNER:DEFAULT_COW, SPAWNER:COAL_COW, SPAWNER:IRON_COW, SPAWNER:GOLD_COW, SPAWNER:DIAMOND_COW`
I then registered custom prices for these spawners:
`'SPAWNER:DEFAULT_COW': 10000
'SPAWNER:COAL_COW': 20000
'SPAWNER:IRON_COW': 30000
'SPAWNER:GOLD_COW': 40000
'SPAWNER:DIAMOND_COW': 50000`
I placed each spawner on my island, ran /is admin level, and I saw 4 different types (I didn't place the non-upgraded spawner) in the /is counts, and my worth value was fine as well.

SpawnerKeyProvider: https://pastebin.com/zLJvKbb6
Main Class: https://pastebin.com/Xa37JVJZ

Let me know if this setup works for you.
